go/arrow/array/list.go

```diff
@@ -618,19 +634,1061 @@ func (b *baseListBuilder) UnmarshalJSON(data []byte) error {
+	for slot := int64(0); slot < int64(a.Len()); slot += 1 {
+		size := l.sizeAt(slot)
+		if size > 0 {
+			offset := l.offsetAt(slot)
+			if offset < 0 || offset > offsetLimit {
+				return outOfBoundsListViewOffset(l, slot, offsetLimit)
+			}
+			if size > offsetLimit-int64(offset) {
+				return outOfBoundsListViewSize(l, slot, offsetLimit)
+			}
+		} else if size < 0 {
+			return outOfBoundsListViewSize(l, slot, offsetLimit)
+		}
+	}
+
+	return nil
+}
+
+func (a *array) validateOffsetsAndMaybeSizes(l offsetsAndSizes, offsetByteWidth int, isListView bool, offsetLimit int64, fullValidation bool) error {
+	nonEmpty := a.Len() > 0
+	if a.data.buffers[1] == nil {
+		// For length 0, an empty offsets buffer is accepted (ARROW-544).
+		if nonEmpty {
+			return fmt.Errorf("non-empty array but offsets are null")
+		} else {
+			return nil
+		}
+	}
+	if isListView {
+		if a.data.buffers[2] == nil {
+			if nonEmpty {
+				return fmt.Errorf("non-empty array but sizes are null")
+			} else {
+				return nil
+			}
+		}
+	}
+
+	var requiredOffsets int
+	if nonEmpty {
+		requiredOffsets = a.Len() + a.Offset()
+		if !isListView {
+			requiredOffsets += 1
+		}
+	} else {
+		requiredOffsets = 0
+	}
+	offsetsByteSize := a.data.buffers[1].Len()
+	if offsetsByteSize/offsetByteWidth < requiredOffsets {
+		return fmt.Errorf("offsets buffer size (bytes): %d isn't large enough for length: %d and offset: %d",
+			offsetsByteSize, a.Len(), a.Offset())
+	}
+	if isListView {
+		requiredSizes := a.Len() + a.Offset()
+		sizesBytesSize := a.data.buffers[2].Len()
+		if sizesBytesSize/offsetByteWidth < requiredSizes {
+			return fmt.Errorf("sizes buffer size (bytes): %d isn't large enough for length: %d and offset: %d",
+				sizesBytesSize, a.Len(), a.Offset())
+		}
+	}
+
+	if fullValidation && requiredOffsets > 0 {
+		if isListView {
+			return a.fullyValidateOffsetsAndSizes(l, offsetLimit)
+		} else {
+			// TODO: implement validation of List and LargeList
+			// return fullyValidateOffsets(offset_limit)
+			return nil
+		}
+	}
+	return nil
+}
+
+func (a *ListView) validate(fullValidation bool) error {
+	values := a.array.data.childData[0]
+	offsetLimit := values.Len()
+	return a.array.validateOffsetsAndMaybeSizes(a, 4, true, int64(offsetLimit), fullValidation)
+}
+
+func (a *ListView) Validate() error {
+	return a.validate(false)
+}
+
+func (a *ListView) ValidateFull() error {
+	return a.validate(true)
+}
+
+func (a *LargeListView) validate(fullValidation bool) error {
+	values := a.array.data.childData[0]
+	offsetLimit := values.Len()
+	return a.array.validateOffsetsAndMaybeSizes(a, 8, true, int64(offsetLimit), fullValidation)
+}
+
+func (a *LargeListView) Validate() error {
+	return a.validate(false)
+}
+
+func (a *LargeListView) ValidateFull() error {
+	return a.validate(true)
+}
+
+type baseListViewBuilder struct {
+	builder
+
+	values  Builder // value builder for the list-view's elements.
+	offsets Builder
+	sizes   Builder
```

Comment:
Don't know why i didn't mention this before, any particular reason you're using `Builder` instead of just a `bufferBuilder`? It seems overkill to have full array builders for these when we just need to build a buffer for each, not entire arrays

Reply:
Cargo-culted from the the `baseListBuilder`.

Reply:
If it can be done quickly/easily I'd prefer for us to use the buffer builders instead, but if it would be more involved to do so then I'm fine with pushing that to a follow up. What do you think?

Reply:
Let me give it a try.

Reply:
Follow-up is better. I just killed my Linux desktop environment because my WIP code contains a memory leak. :sweat:

Reply:
No problem. It's not a correctness issue, just a potential for better performance. So we can do it as a follow-up